rocksdb/src/main/java/com/yahoo/ycsb/db/rocksdb/RocksDBClient.java

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) 2017 YCSB contributors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package com.yahoo.ycsb.db.rocksdb;
+
+import com.yahoo.ycsb.*;
+import com.yahoo.ycsb.Status;
+import net.jcip.annotations.GuardedBy;
+import org.rocksdb.*;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.file.*;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * RocksDB binding for <a href="http://rocksdb.org/">RocksDB</a>.
+ *
+ * See {@code rocksdb/README.md} for details.
+ *
+ * @author <a href="mailto:adam@evolvedbinary.com">Adam Retter</a>
+ */
+public class RocksDBClient extends DB {
+
+  public static final String PROP_ROCKSDB_DIR = "rocksdb.dir";
+  private static final String CF_NAMES_FILENAME = "CF_NAMES";
+
+  @GuardedBy("RocksDBClient.class") private static Path rocksdbDir = null;
+  @GuardedBy("RocksDBClient.class") private static RocksObject dbOptions = null;
+  @GuardedBy("RocksDBClient.class") private static RocksDB rocksDb = null;
```

Comment:
if this wasn't static, you'd not need all the synchronization.... What is the advantage of forcing the `RocksDB` object you're using to be a singleton?

Reply:
I wanted to ensure that there was only a single instance of RocksDB running at any time within YCSB.  Perhaps this is unnecessary?

Reply:
I get that you want only a single instance of RocksDB. I don't understand **why**. For the general case, that is unnecessary, yes. It may be different for RocksDB though. I have never worked with it before...

So ... why do you want to ensure that there is only a single instance of RocksDB running at any time?
How do you propose dealing with multiple ycsb run invocations on the same host?

Reply:
RocksDB is an embedded database (not a database server), i.e. it will be running inside the same JVM as YCSB itself. I wanted to ensure there was only one RocksDB instance per YCSB instance. Does that make sense?